File: annotation_adience_keras.pl
#Generate annotation for keras
#https://www.openu.ac.il/home/hassner/Adience/data.html#agegender

use warnings;
use strict;
use Image::Size;
use File::Copy;

my $input_path="dataset/adience/";
my $dataset_path="dataset/agegender_adience/";

my $ANNOTATION_FILES="./$input_path"."fold_";
my $FACE_FILES="./$input_path"."aligned";

mkdir "$dataset_path"."annotations/agegender/";
mkdir "$dataset_path"."annotations/agegender/validation";
mkdir "$dataset_path"."annotations/agegender/train";

mkdir "$dataset_path"."annotations/gender";
mkdir "$dataset_path"."annotations/gender";
mkdir "$dataset_path"."annotations/gender/validation";
mkdir "$dataset_path"."annotations/gender/train";

mkdir "$dataset_path"."annotations/age";
mkdir "$dataset_path"."annotations/age";
mkdir "$dataset_path"."annotations/age/validation";
mkdir "$dataset_path"."annotations/age/train";

my $file_no=0;
my $line_no=0;

my $imagew;
my $imageh;

my $before_file="";

for(my $list=0;$list<5;$list=$list+1){
  open(IN,"$ANNOTATION_FILES"."$list"."_data.txt") or die ("agegender dataset not found");

  my $header=<IN>;

  while(my $line=<IN>){
    #print $line;

    my $user_id;
    my $original_image;
    my $face_id;
    my $age;
    my $gender;
    my $x;
    my $y;
    my $dx;
    my $dy;
    my $tilt_ang;
    my $fiducial_yaw_angle;
    my $fiducial_score;

    ($user_id,$original_image,$face_id,$age,$gender,$x,$y,$dx,$dy,$tilt_ang,$fiducial_yaw_angle,$fiducial_score)=split("\t",$line);

    $x=$x+$dx/2;
    $y=$y+$dy/2;
    my $w=$dx;
    my $h=$dy;
    
    my $category=-1;
    my $category_label="";
    my $age_int=-1;
    if($age =~ /^[0-9]+$/){
      $age_int=int($age);
    }else{
      if($age =~ /\(([0-9]+),/){
        $age_int=int("$1");
      }
    }

    if($age_int>=0 and $age_int<=3){
      $category=0;
      $category_label="0-2";
    }
    if($age_int>=4 && $age_int<=7){
      $category=1;
      $category_label="4-6";
    }
    if($age_int>=8 && $age_int<=14){
      $category=2;
      $category_label="8-13";
    }
    if($age_int>=15 && $age_int<=24){
      $category=3;
      $category_label="15-20";
    }
    if($age_int>=25 && $age_int<=37){
      $category=4;
      $category_label="25-32";
    }
    if($age_int>=38 && $age_int<=47){
      $category=5;
      $category_label="38-43";
    }
    if($age_int>=48 && $age_int<=59){
      $category=6;
      $category_label="48-53";
    }
    if($age_int>=60 && $age_int<=100){
      $category=7;
      $category_label="60-";
    }
    if($age eq "None"){
      #skip unknown age from dataset
      next;
    }
    if($category eq -1){
      print "Age not found $age\n$line\n";
      exit;
    }

    if($gender eq "f"){
      $category=$category+8;
    }else{
      if($gender eq "m"){
        $category=$category+0;
      }else{
        #print "Gender not found $gender\n$line\n";
        next;
      }
    }

    my $label="$category"."_"."$category_label"."_"."$gender";
    my $label_age="$category_label";
    my $label_gender="$gender";

    my $thumb_dir="$FACE_FILES/$user_id/";

    opendir(THUMB, "$thumb_dir") or die "usage: $0 thumb_dir\n";
    my $filepath="";
    foreach my $dir (readdir(THUMB)) {
      next if ($dir eq '.' || $dir eq '..');
      next if ($dir eq '.DS_Store');
      if($dir =~ /$face_id\.$original_image/){
        $filepath=$dir;
        last;
      }
    }

    if($filepath eq ""){
      print "image file not found\n";
      next;
    }

    mkdir "./$dataset_path"."annotations/agegender/validation/$label";
    mkdir "./$dataset_path"."annotations/agegender/train/$label";

    mkdir "./$dataset_path"."annotations/age/validation/$label_age";
    mkdir "./$dataset_path"."annotations/age/train/$label_age";

    mkdir "./$dataset_path"."annotations/gender/validation/$label_gender";
    mkdir "./$dataset_path"."annotations/gender/train/$label_gender";

    if($before_file ne $original_image){
      #if($line_no ne 0){
      #  close(OUT);
      #  $line_no=0;
      #}
      $before_file=$original_image;
      #$line_no=1;
      ($imagew, $imageh) = imgsize("$FACE_FILES/$user_id/$filepath");

      $file_no=$file_no+1;
      if($file_no%4 eq 0){
        copy("$FACE_FILES/$user_id/$filepath","$dataset_path"."annotations/agegender/validation/$label/$filepath");
        copy("$FACE_FILES/$user_id/$filepath","$dataset_path"."annotations/age/validation/$label_age/$filepath");
        copy("$FACE_FILES/$user_id/$filepath","$dataset_path"."annotations/gender/validation/$label_gender/$filepath");
      }else{
        copy("$FACE_FILES/$user_id/$filepath","$dataset_path"."annotations/agegender/train/$label/$filepath");
        copy("$FACE_FILES/$user_id/$filepath","$dataset_path"."annotations/age/train/$label_age/$filepath");
        copy("$FACE_FILES/$user_id/$filepath","$dataset_path"."annotations/gender/train/$label_gender/$filepath");
      }
    }
    
    $x=1.0*$x/$imagew;
    $y=1.0*$y/$imagew;
    $w=1.0*$w/$imagew;
    $h=1.0*$h/$imagew;
    #print OUT "$category $x $y $w $h\n";
  }

  #if($line_no ne 0){
  #  close(OUT);
  #}

  close(IN);
}

File: annotation_adience_keras.py
#Generate annotation for keras
#https://www.openu.ac.il/home/hassner/Adience/data.html#agegender

from scipy import io as spio
from datetime import datetime

import re
import numpy as np
import os
import shutil
import sys
import glob

if(os.path.exists("./dataset/imdb_crop/")):
	DATASET_ROOT_PATH=""
else:
	DATASET_ROOT_PATH="/Volumes/ST5/Keras/"

OUTPUT_LABEL="agegender_adience"

ANNOTATION_FILES=DATASET_ROOT_PATH+"dataset/adience/fold_"
FACE_FILES=DATASET_ROOT_PATH+"dataset/adience/aligned"

GENDER_PATH=DATASET_ROOT_PATH+"dataset/"+OUTPUT_LABEL+"/annotations/gender/"
AGE_PATH=DATASET_ROOT_PATH+"dataset/"+OUTPUT_LABEL+"/annotations/age/"
AGE101_PATH=DATASET_ROOT_PATH+"dataset/"+OUTPUT_LABEL+"/annotations/age101/"

if(not os.path.exists(DATASET_ROOT_PATH+"dataset/"+OUTPUT_LABEL+"/")):
	os.mkdir(DATASET_ROOT_PATH+"dataset/"+OUTPUT_LABEL)
if(not os.path.exists(DATASET_ROOT_PATH+"dataset/"+OUTPUT_LABEL+"/annotations/")):
	os.mkdir(DATASET_ROOT_PATH+"dataset/"+OUTPUT_LABEL+"/annotations")

if(not os.path.exists(GENDER_PATH)):
	os.mkdir(GENDER_PATH)
	os.mkdir(GENDER_PATH+"train")
	os.mkdir(GENDER_PATH+"train/f")
	os.mkdir(GENDER_PATH+"train/m")
	os.mkdir(GENDER_PATH+"validation")
	os.mkdir(GENDER_PATH+"validation/f")
	os.mkdir(GENDER_PATH+"validation/m")

if(not os.path.exists(AGE_PATH)):
	os.mkdir(AGE_PATH)
	os.mkdir(AGE_PATH+"train")
	os.mkdir(AGE_PATH+"train/0-2")
	os.mkdir(AGE_PATH+"train/4-6")
	os.mkdir(AGE_PATH+"train/8-13")
	os.mkdir(AGE_PATH+"train/15-20")
	os.mkdir(AGE_PATH+"train/25-32")
	os.mkdir(AGE_PATH+"train/38-43")
	os.mkdir(AGE_PATH+"train/48-53")
	os.mkdir(AGE_PATH+"train/60-")
	os.mkdir(AGE_PATH+"validation")
	os.mkdir(AGE_PATH+"validation/0-2")
	os.mkdir(AGE_PATH+"validation/4-6")
	os.mkdir(AGE_PATH+"validation/8-13")
	os.mkdir(AGE_PATH+"validation/15-20")
	os.mkdir(AGE_PATH+"validation/25-32")
	os.mkdir(AGE_PATH+"validation/38-43")
	os.mkdir(AGE_PATH+"validation/48-53")
	os.mkdir(AGE_PATH+"validation/60-")

if(not os.path.exists(AGE101_PATH)):
	os.mkdir(AGE101_PATH)
	os.mkdir(AGE101_PATH+"train")
	os.mkdir(AGE101_PATH+"validation")
	for i in range(0,101):
		os.mkdir(AGE101_PATH+"train/"+format(i, '03d'))
		os.mkdir(AGE101_PATH+"validation/"+format(i, '03d'))

def get_gender_path(gender):
	if(gender=="f"):
		return "f"
	return "m"

def get_age_path(age):
	if(age>=0 and age<=3):
		return "0-2"
	if(age>=4 and age<=7):
		return "4-6"
	if(age>=8 and age<=14):
		return "8-13"
	if(age>=15 and age<=24):
		return "15-20"
	if(age>=25 and age<=37):
		return "25-32"
	if(age>=38 and age<=47):
		return "38-43"
	if(age>=48 and age<=59):
		return "48-53"
	return "60-"

i=0

for list in range(5):
	path=ANNOTATION_FILES+str(list)+"_data.txt"
	lines=open(path).readlines()
	for line in lines:
		data=line.split("\t");
		user_id,original_image,face_id,age,gender,x,y,dx,dy,tilt_ang,fiducial_yaw_angle,fiducial_score=data
		
		if age=="age":
			continue

		result = re.match(r"^[0-9]+$",age)
		if result:
			age_int=int(age)
		else:
			result = re.match(r"\(([0-9]+),",age)
			if result:
				age_int=int(result.group(1))
			else:
				age_int=-1

		age=age_int

		thumb_dir=FACE_FILES+"/"+user_id+"/";
		full_path=""
		for image_path in glob.glob(thumb_dir+"*"):
			image_name = os.path.basename(image_path)
			if image_name.find(face_id+"."+original_image)!=-1:
				full_path=image_path
				break

		if full_path=="":
			print("path not found")
			sys.exit(1)

		print("path:"+str(full_path)+" gender:"+str(gender))
		print(""+get_age_path(age)+" "+get_gender_path(gender))
		train_or_validation="train"
		if(i%4==0):
			train_or_validation="validation"
		src_img=full_path
		if age_int!=-1:
			shutil.copyfile(src_img, AGE_PATH+train_or_validation+"/"+get_age_path(age)+"/"+str(i)+".jpg")
		shutil.copyfile(src_img, GENDER_PATH+train_or_validation+"/"+get_gender_path(gender)+"/"+str(i)+".jpg")
		if age_int!=-1:
			shutil.copyfile(src_img, AGE101_PATH+train_or_validation+"/"+format(age,"03d")+"/"+str(i)+".jpg")

		i=i+1